org.eclipse.xtext.util/src/org/eclipse/xtext/util/XtextVersion.java

```diff
@@ -92,24 +82,73 @@ public String toString() {
+	 * by "-SNAPSHOT". Otherwise it will return the base version without the qualifier.
+	 * 
+	 * @return Version string in format <code>MAJOR.MINOR.MICRO[-SNAPSHOT]</code>
+	 */
 	private static String readVersionFromManifest() {
-		URL baseURL = XtextVersion.Plugin.INSTANCE.getBaseURL();
-		try (InputStream is = new URL (baseURL + "META-INF/MANIFEST.MF").openStream()) {
-			Manifest manifest = new Manifest(is);
-			String version = manifest.getMainAttributes().getValue("Maven-Version");
-			if ("unspecified".equals(version)) {
-				version = manifest.getMainAttributes().getValue("Bundle-Version");
-				if (version.endsWith(".qualifier")) {
-					return version.replace(".qualifier", "-SNAPSHOT");
-				} else {
-					return version.substring(0, version.lastIndexOf("."));
+		try {
+			URL ownManifestURL = getOwnManifestURL();
+			
+			try (InputStream is = ownManifestURL.openStream()) {
+				Manifest manifest = new Manifest(is);
+				String version = manifest.getMainAttributes().getValue("Maven-Version");
+				if ("unspecified".equals(version)) {
+					version = manifest.getMainAttributes().getValue("Bundle-Version");
+					if (version.endsWith(".qualifier")) {
+						return version.replace(".qualifier", "-SNAPSHOT");
+					} else {
+						return version.substring(0, version.lastIndexOf("."));
+					}
 				}
+				return version;
 			}
-			return version;
 		} catch (IOException e) {
 			throw new RuntimeException(e);
 		}
 	}
+	
+	/**
+	 * Locates the URL of this bundle's manifest file. This is done by locating the manifest file
+	 * relative to the class resource of this class. The location can be either be a file URL in the context
+	 * of unit test execution or an archive URL for CI/release bundle.
+	 * 
+	 * @throws FileNotFoundException When the MANIFEST.MF for org.eclipse.xtext.util was not found
+	 */
+	private static URL getOwnManifestURL () throws FileNotFoundException {
+		// can't use XtextVersion.class.getResource() as it retrieves the first manifest found,
+		// not the one for this bundle. Thus we have to search for the right one first.
+		// we do this here by locating the class file of XtextVersion from its classloader and locate
+		// the manifest relative to the base URI
+		try {
+			String ownClassAsResourcePath = XtextVersion.class.getName().replace('.', '/')+".class";
+			URL ownClassResourceURL = XtextVersion.class.getClassLoader().getResource(ownClassAsResourcePath);
+			URI ownClassResourceURI = URI.createURI(ownClassResourceURL.toURI().toString());
+			
+			// The XtextVersion class has 5 segments. When the class is bundled in a jar the manifest is located 
+			// relative to the root.
+			// In a unit test the class is located in the build output directory, but the manifest
+			// is found relative to the project root. This can be either 1 or 2 levels above (bin/main in Gradle or target in Maven).
+			for (int i=5; i<=7; i++) {
```

Comment:
am not sure if i like this.

Reply:
Could check for file or archive URLs. But how to make the URL reliable relative to the base URL? I agree it does not look too nice, but I did not came to a better solution.